File: 525_contiguous_array/main.py
class Solution:
    def findMaxLength(self, nums: List[int]) -> int:
        nums[0] = 1 if nums[0] == 1 else -1
        for i in range(1, len(nums)):
            nums[i] = nums[i - 1] + (1 if nums[i] == 1 else -1)
    
        d = {}
        d[0] = -1
        result = 0
        for i in range(len(nums)):
            if nums[i] in d:
                result = max(result, i - d[nums[i]])
            else:
                d[nums[i]] = i
        
        return result

File: 525_contiguous_array/main.java
class Solution {
    public int findMaxLength(int[] nums) {
        int[] cumulative = new int[nums.length + 1];
        cumulative[0] = 0;
        for (int i = 0; i < nums.length; ++i) {
            if (nums[i] == 1) {
                cumulative[i+1] = cumulative[i] + 1;
            } else {
                cumulative[i+1] = cumulative[i] - 1;
            }
        }

        HashMap<Integer, Integer> found = new HashMap<>();
        int result = 0;
        for (int i = 0; i < cumulative.length; ++i) {
            if (found.containsKey(cumulative[i])) {
                result = Math.max(result, i - found.get(cumulative[i]));
            } else {
                found.put(cumulative[i], i);
            }
        }
        
        return result;
    }
}
